source/slang/slang-ir-sccp.cpp

```diff
@@ -288,10 +793,116 @@ struct SCCPContext
+                getLatticeVal(inst->getOperand(1)));
+        case kIROp_Greater:
+            return evalGreater(
+                inst->getFullType(),
+                getLatticeVal(inst->getOperand(0)),
+                getLatticeVal(inst->getOperand(1)));
+        case kIROp_Less:
+            return evalLess(
+                inst->getFullType(),
+                getLatticeVal(inst->getOperand(0)),
+                getLatticeVal(inst->getOperand(1)));
+        case kIROp_Leq:
+            return evalLeq(
+                inst->getFullType(),
+                getLatticeVal(inst->getOperand(0)),
+                getLatticeVal(inst->getOperand(1)));
+        case kIROp_Geq:
+            return evalGeq(
+                inst->getFullType(),
+                getLatticeVal(inst->getOperand(0)),
+                getLatticeVal(inst->getOperand(1)));
+        case kIROp_And:
+            return evalAnd(
+                inst->getFullType(),
+                getLatticeVal(inst->getOperand(0)),
+                getLatticeVal(inst->getOperand(1)));
+        case kIROp_Or:
+            return evalOr(
+                inst->getFullType(),
+                getLatticeVal(inst->getOperand(0)),
+                getLatticeVal(inst->getOperand(1)));
+        case kIROp_Not:
+            return evalNot(inst->getFullType(), getLatticeVal(inst->getOperand(0)));
+        case kIROp_BitAnd:
+            return evalBitAnd(
+                inst->getFullType(),
+                getLatticeVal(inst->getOperand(0)),
+                getLatticeVal(inst->getOperand(1)));
+        case kIROp_BitOr:
+            return evalBitOr(
+                inst->getFullType(),
+                getLatticeVal(inst->getOperand(0)),
+                getLatticeVal(inst->getOperand(1)));
+        case kIROp_BitNot:
+            return evalBitNot(inst->getFullType(), getLatticeVal(inst->getOperand(0)));
+        case kIROp_BitXor:
+            return evalBitXor(
+                inst->getFullType(),
+                getLatticeVal(inst->getOperand(0)),
+                getLatticeVal(inst->getOperand(1)));
+        case kIROp_BitCast:
+            return evalBitCast(inst->getFullType(), getLatticeVal(inst->getOperand(0)));
+        case kIROp_Neg:
+            return evalNeg(inst->getFullType(), getLatticeVal(inst->getOperand(0)));
+        case kIROp_Lsh:
+            return evalLsh(
+                inst->getFullType(),
+                getLatticeVal(inst->getOperand(0)),
+                getLatticeVal(inst->getOperand(1)));
+        case kIROp_Rsh:
+            return evalRsh(
+                inst->getFullType(),
+                getLatticeVal(inst->getOperand(0)),
+                getLatticeVal(inst->getOperand(1)));
+        case kIROp_Select:
+            return evalSelect(
+                getLatticeVal(inst->getOperand(0)),
+                getLatticeVal(inst->getOperand(1)),
+                getLatticeVal(inst->getOperand(2)));
+        default:
+            break;
```

Comment:
This all seems good, and we can keep adding to it. I think there are opportunities to factor the folding and simplification logic to be more streamlined, and usable from outside the SCCP pass if we ever want that. E.g., it would be really nice if we could have access to these kind of simplifications as part of the `IRBuilder`, since it could mean that in truly simple cases we simply don't even emit any IR for these operations.